src/rlx_resolve.erl

```diff
@@ -95,46 +97,97 @@ find_and_remove(ExcludeName, [#{name := Name} | Rest]) when ExcludeName =:= Name
 find_and_remove(ExcludeName, [H | Rest]) ->
     [H | find_and_remove(ExcludeName, Rest)].
 
-find_app(Name, Vsn, []) ->
+%% Applications are first searched for in the `Apps' variable which is a map
+%% of application name to `rlx_app_info' map. This variable is passed to `relx'
+%% function calls to build the release or tarball
+%%
+%% If the application isn't found in the `Apps' map then the `lib_dirs' list of
+%% directories is searched for the application under `<dir>/*/ebin/Name.app'.
+%% Lastly, if the application still isn't found then the code path is checked
+%% using `code:lib_dir'.
+find_app(Name, Vsn, Apps, LibDirs) ->
+    case maps:find(Name, Apps) of
+        {ok, AppInfo} ->
+            %% verify the app is the version we want and if not try
+            %% finding the app-vsn needed in the configured paths
+            case check_app(Name, Vsn, AppInfo) of
+                true ->
+                    AppInfo;
+                false ->
+                    search_for_app(Name, Vsn, LibDirs)
+            end;
+        error ->
+            search_for_app(Name, Vsn, LibDirs)
+    end.
+
+search_for_app(Name, Vsn, LibDirs) ->
+    case find_app_in_dir(Name, Vsn, LibDirs) of
+        not_found ->
+            find_app_in_code_path(Name, Vsn);
+        AppInfo ->
+            case check_app(Name, Vsn, AppInfo) of
+                true ->
+                    AppInfo;
+                false ->
+                    find_app_in_code_path(Name, Vsn)
+            end
+    end.
+
+find_app_in_dir(_Name, _Vsn, []) ->
+    not_found;
+find_app_in_dir(Name, Vsn, [Dir | Rest]) ->
+    AppFile = filename:join([Dir, "*", "ebin", [Name, ".app"]]),
+    case filelib:wildcard(AppFile) of
+        [] ->
+            find_app_in_dir(Name, Vsn, Rest);
+        Matches ->
+            case rlx_util:list_search(fun(AppFilePath) ->
+                                              to_app(Name, Vsn, AppFilePath)
+                                      end, Matches) of
+                {value, App} ->
+                    App;
+                false ->
+                    find_app_in_dir(Name, Vsn, Rest)
+            end
+    end.
+
+find_app_in_code_path(Name, Vsn) ->
     case code:lib_dir(Name) of
         {error, bad_name} ->
             erlang:error(?RLX_ERROR({app_not_found, Name}));
         Dir ->
-            to_app(Name, Vsn, Dir)
-    end;
-find_app(Name, Vsn, [App | Rest]) ->
-    case check_app(Name, Vsn, App) of
-        true ->
-            App;
-        false ->
-            find_app(Name, Vsn, Rest)
+            case to_app(Name, Vsn, filename:join([Dir, "ebin", [Name, ".app"]])) of
+                {true, AppInfo} ->
+                    AppInfo;
+                false ->
+                    erlang:error(?RLX_ERROR({app_not_found, Name}))
```

Comment:
Am I right in thinking that this bit of code assumes that in code:lib_dir, there is only going to be one app version present only for each app? Because I think the code module only sees one of them, as compared to the lib_dir lookup you have prior to that one.

Reply:
Yes, that is how the code module works, right? So it is the only option for code path.